algorithms/nautilus/fuzzer/state.cpp

```diff
@@ -0,0 +1,469 @@
+        /* Print testcase because we don't want to lose it */
+        std::cout << buffer << std::endl;
+        throw exceptions::unable_to_create_file(
+          Util::StrPrintf("Cannot save crash: %s", filepath.c_str()),
+          __FILE__, __LINE__
+        );
+      }
+      Util::WriteFile(fd, buffer.data(), buffer.size());
+      Util::CloseFile(fd);
+
+      break;
+    }
+
+    default: // pass
+      break;
+  }
+  // exit_reason
+  exec_reason = exec_reason;
+}
+
+/**
+ * @fn
+ * @brief Execute PUT with raw data as input
+ * @param (code) Testcase
+ * @return Execution time in nanoseconds
+ */
+u64 NautilusState::ExecRaw(const std::string& code) {
+  using namespace std::chrono;
+
+  execution_count++;
+
+  system_clock::time_point start, end;
+  const u8* data = reinterpret_cast<const u8*>(code.data());
+  u32 size = code.size();
+
+  if (u32 tmout = setting->exec_timeout_ms == 0) {
+    /* Run without timeout */
+    start = system_clock::now();
+    executor->Run(data, size);
+    end = system_clock::now();
+  } else {
+    /* Run with timeout*/
+    start = system_clock::now();
+    executor->Run(data, size, tmout);
+    end = system_clock::now();
+  }
+
+  /* Calculate average execution */
+  u64 execution_time = duration_cast<nanoseconds>(end - start).count();
+  this->average_executions_per_sec = this->average_executions_per_sec * 0.9 \
+    + ((1.0 / static_cast<double>(execution_time)) * 1000000000.0) * 0.1;
+
+  return execution_time;
+}
+
+/**
+ * @fn
+ * @brief Update bits by checking deterministic behavior
+ */
+void NautilusState::CheckForDeterministicBehavior(
+  const std::vector<uint8_t>& old_bitmap,
+  std::vector<size_t>& new_bits,
+  const std::string& code
+) {
+  for (size_t i = 0; i < 5; i++) {
+    ExecRaw(code);
+    InplaceMemoryFeedback new_feedback = executor->GetAFLFeedback();
+
+    new_feedback.ShowMemoryToFunc(
+      [&old_bitmap, &new_bits](const u8* run_bitmap, u32) {
+        /* This loop looks unnecessary
+        for (size_t j = 0; j < len; j++) {
+          if (run_bitmap[j] != old_bitmap.at(j)) {
+            std::cout << "[-] Found fucky bit " << j << std::endl;
+          }
+        }
+        */
+
+        /* Retain bits */
+        for (ssize_t j = new_bits.size() - 1; j >= 0; j--) {
+          if (run_bitmap[new_bits[j]] == 0) {
+            new_bits.erase(new_bits.begin() + j);
+          }
+        }
+      }
+    );
+  }
+}
+
+bool NautilusState::HasBits(const TreeLike& tree,
+                            std::unordered_set<size_t>& bits,
+                            ExecutionReason exec_reason,
+                            Context& ctx) {
+  RunOnWithoutDedup(tree, exec_reason, ctx);
+
+  InplaceMemoryFeedback new_feedback = executor->GetAFLFeedback();
+
+  bool found_all = true;
+  new_feedback.ShowMemoryToFunc(
+    [&bits, &found_all](const u8* run_bitmap, u32 len) {
+      for (size_t bit: bits) {
+        // TODO: remove this assert
+        assert (bit < len);
```

Comment:
Please use DEBUG_ASSERT and remove assert in every place

Reply:
Modified in [d79f096](https://github.com/fuzzuf/fuzzuf/pull/23/commits/d79f096c2946eb1b5c26159d1acc2f3ab5616b40).
This commit also introduces a new macro `UNUSED`:
https://github.com/fuzzuf/fuzzuf/blob/d79f096c2946eb1b5c26159d1acc2f3ab5616b40/include/fuzzuf/utils/common.hpp#L107
e,g,.
https://github.com/fuzzuf/fuzzuf/blob/d79f096c2946eb1b5c26159d1acc2f3ab5616b40/algorithms/nautilus/grammartec/tree.cpp#L563-L564